sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/CSharpDiagnosticAnalyzerContextHelper.cs

```diff
@@ -115,10 +115,11 @@ internal static class CSharpDiagnosticAnalyzerContextHelper
 
         #region SyntaxTree.IsGenerated
 
-        internal static bool IsGenerated(this SyntaxTree tree,
+        internal static bool ShouldAnalyze(this SyntaxTree tree, AnalyzerOptions options,
             Compilation compilation)
         {
-            return tree.IsGenerated(CSharp.CSharpGeneratedCodeRecognizer.Instance, compilation);
+            return !tree.IsGenerated(CSharp.CSharpGeneratedCodeRecognizer.Instance, compilation) ||
+                options.ShouldAnalyzeGeneratedCode(LanguageNames.CSharp);
```

Comment:
TODO: once the options check has been optimized, it would be better to do the options check first as it is a global check rather than a per-file check. It would also be worth recording this in a comment e.g.

` // PERF: do the global setting check before the per-file check. `

Reply:
Also, couldn't this method call `DiagnosticAnalyzerContextHelper.ShouldAnalyze` instead of repeating the same checks?